src/lib/core/tests/TestTLVVectorWriter.cpp

```diff
@@ -0,0 +1,180 @@
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <lib/core/TLVVectorWriter.h>
+
+#include <cstddef>
+#include <cstdint>
+#include <string>
+#include <vector>
+
+#include <lib/core/CHIPError.h>
+#include <lib/core/ErrorStr.h>
+#include <lib/core/TLVCommon.h>
+#include <lib/core/TLVTags.h>
+#include <lib/support/Span.h>
+#include <lib/support/UnitTestContext.h>
+#include <lib/support/UnitTestExtendedAssertions.h>
+#include <lib/support/UnitTestRegistration.h>
+#include <lib/support/UnitTestUtils.h>
+
+using namespace chip;
+using namespace chip::TLV;
+
+/**
+ * context
+ */
+
+struct TestTLVContext
+{
+    nlTestSuite * mSuite   = nullptr;
+    int mEvictionCount     = 0;
+    uint32_t mEvictedBytes = 0;
+
+    TestTLVContext(nlTestSuite * suite) : mSuite(suite) {}
+};
+
+void InitAndFinalizeWithNoData(nlTestSuite * inSuite, void * inContext)
+{
+    std::vector<uint8_t> buffer;
+    TlvVectorWriter writer(&buffer);
+
+    // Init and finalize but write not data.
+    NL_TEST_ASSERT(inSuite, writer.Finalize() == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, buffer.empty());
+}
+
+void SingleSmallDataFitsInOriginalBuffer(nlTestSuite * inSuite, void * inContext)
+{
+    std::vector<uint8_t> buffer;
+    TlvVectorWriter writer(&buffer);
+    TLVReader reader;
+
+    NL_TEST_ASSERT(inSuite, writer.Put(AnonymousTag(), true) == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, writer.Finalize() == CHIP_NO_ERROR);
+
+    reader.Init(buffer.data(), buffer.size());
+    NL_TEST_ASSERT(inSuite, reader.Next() == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, reader.GetTag() == AnonymousTag());
+
+    bool value = false;
+    NL_TEST_ASSERT(inSuite, reader.Get(value) == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, value == true);
+}
+
+void SingleLargeDataRequiresNewBufferAllocation(nlTestSuite * inSuite, void * inContext)
+{
+    std::vector<uint8_t> buffer;
+    TlvVectorWriter writer(&buffer);
+    TLVReader reader;
+    static constexpr size_t kStringSize = 10000;
+
+    const std::string bytes(kStringSize, 'a');
+    CHIP_ERROR error = writer.PutString(AnonymousTag(), bytes.data());
+    NL_TEST_ASSERT(inSuite, error == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, writer.Finalize() == CHIP_NO_ERROR);
+
+    reader.Init(buffer.data(), buffer.size());
+    NL_TEST_ASSERT(inSuite, reader.Next() == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, reader.GetTag() == AnonymousTag());
+
+    CharSpan span;
+    error = reader.Get(span);
+    NL_TEST_ASSERT(inSuite, error == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, std::string(span.data(), span.size()) == bytes);
+}
+
+void ManySmallDataRequiresNewBufferAllocation(nlTestSuite * inSuite, void * inContext)
```

Comment:
Is it possible to add a test to allocate a PacketBuffer(from heap) after a large TLV vector is populated and copy the contents of the vector into the allocated PacketBuffer?

Reply:
We should not have this scenario, as the PacketBuffer is typically used for content transmission on the wire, while the TlvVectorBuffer is used for creating the TLV binary format to be stored within the PacketBuffer.